File: src/sokoenginepy/common/hexoban_tessellation.py
from __future__ import annotations

from typing import Dict, Final, Tuple

from .characters import Characters
from .config import Config
from .direction import Direction
from .tessellation import index_1d, index_column, index_row, is_on_board_2d
from .tessellation_impl import PusherStepData, TessellationImpl


class HexobanTessellation(TessellationImpl):
    _LEGAL_DIRECTIONS: Final[Tuple[Direction, ...]] = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    _CHR_TO_PUSHER_STEP: Final[Dict[str, PusherStepData]] = {
        Characters.l: PusherStepData(Direction.LEFT, False),
        Characters.L: PusherStepData(Direction.LEFT, True),
        Characters.r: PusherStepData(Direction.RIGHT, False),
        Characters.R: PusherStepData(Direction.RIGHT, True),
        Characters.u: PusherStepData(Direction.NORTH_WEST, False),
        Characters.U: PusherStepData(Direction.NORTH_WEST, True),
        Characters.d: PusherStepData(Direction.SOUTH_EAST, False),
        Characters.D: PusherStepData(Direction.SOUTH_EAST, True),
        Characters.n: PusherStepData(Direction.NORTH_EAST, False),
        Characters.N: PusherStepData(Direction.NORTH_EAST, True),
        Characters.s: PusherStepData(Direction.SOUTH_WEST, False),
        Characters.S: PusherStepData(Direction.SOUTH_WEST, True),
    }

    _PUSHER_STEP_TO_CHR: Final[Dict[PusherStepData, str]] = {
        v: k for k, v in _CHR_TO_PUSHER_STEP.items()
    }

    def neighbor_position(
        self, position: int, direction: Direction, board_width: int, board_height: int
    ) -> int:
        if position < 0:
            raise IndexError(f"Position {position} is invalid value!")

        if board_width < 0:
            raise ValueError(f"Board width {board_width} is invalid value!")

        if board_height < 0:
            raise ValueError(f"Board height {board_height} is invalid value!")

        row = index_row(position, board_width)
        column = index_column(position, board_width)

        if direction == Direction.LEFT:
            column -= 1
        elif direction == Direction.RIGHT:
            column += 1
        elif direction == Direction.NORTH_EAST:
            column += row % 2
            row -= 1
        elif direction == Direction.NORTH_WEST:
            column -= (row + 1) % 2
            row -= 1
        elif direction == Direction.SOUTH_EAST:
            column += row % 2
            row += 1
        elif direction == Direction.SOUTH_WEST:
            column -= (row + 1) % 2
            row += 1
        else:
            raise ValueError(
                f"Unsupported direction {direction} for {self.__class__.__name__}"
            )

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return Config.NO_POS

File: src/libsokoengine/common/hexoban_tessellation.cpp
/// @file
#include "hexoban_tessellation.hpp"

#include "characters.hpp"
#include "pusher_step.hpp"

using std::invalid_argument;
using std::string;

namespace sokoengine {
namespace implementation {

static const Directions HEX_LEGAL_DIRECTIONS = {
  Direction::LEFT,
  Direction::RIGHT,
  Direction::NORTH_EAST,
  Direction::NORTH_WEST,
  Direction::SOUTH_EAST,
  Direction::SOUTH_WEST};

const Directions &HexobanTessellation::legal_directions() const {
  return HEX_LEGAL_DIRECTIONS;
}

position_t HexobanTessellation::neighbor_position(
  position_t       position,
  const Direction &direction,
  board_size_t     width,
  board_size_t     height
) const {
  position_t row = index_y(position, width), column = index_x(position, width);
  switch (direction) {
    case Direction::LEFT:
      column += -1;
      break;
    case Direction::RIGHT:
      column += 1;
      break;
    case Direction::NORTH_EAST:
      column += row % 2;
      row -= 1;
      break;
    case Direction::NORTH_WEST:
      column -= (row + 1) % 2;
      row -= 1;
      break;
    case Direction::SOUTH_EAST:
      column += row % 2;
      row += 1;
      break;
    case Direction::SOUTH_WEST:
      column -= (row + 1) % 2;
      row += 1;
      break;
    default:
      throw invalid_argument(
        "Unsupported direction " + to_str(direction) + " for HexobanTessellation!"
      );
  }

  if (is_on_board_2d(column, row, width, height))
    return index_1d(column, row, width);
  else
    return Config::NO_POS;
}

PusherStep HexobanTessellation::char_to_pusher_step(char rv) const {
  switch (rv) {
    case Characters::l:
      return PusherStep(Direction::LEFT, Config::NO_ID);
    case Characters::L:
      return PusherStep(Direction::LEFT, Config::DEFAULT_ID);

    case Characters::r:
      return PusherStep(Direction::RIGHT, Config::NO_ID);
    case Characters::R:
      return PusherStep(Direction::RIGHT, Config::DEFAULT_ID);

    case Characters::u:
      return PusherStep(Direction::NORTH_WEST, Config::NO_ID);
    case Characters::U:
      return PusherStep(Direction::NORTH_WEST, Config::DEFAULT_ID);

    case Characters::d:
      return PusherStep(Direction::SOUTH_EAST, Config::NO_ID);
    case Characters::D:
      return PusherStep(Direction::SOUTH_EAST, Config::DEFAULT_ID);

    case Characters::n:
      return PusherStep(Direction::NORTH_EAST, Config::NO_ID);
    case Characters::N:
      return PusherStep(Direction::NORTH_EAST, Config::DEFAULT_ID);

    case Characters::s:
      return PusherStep(Direction::SOUTH_WEST, Config::NO_ID);
    case Characters::S:
      return PusherStep(Direction::SOUTH_WEST, Config::DEFAULT_ID);

    default:
      throw invalid_argument(
        string("Illegal PusherStep character '") + rv + "' in HexobanTessellation!"
      );
  }
}

char HexobanTessellation::pusher_step_to_char(const PusherStep &rv) const {
  switch (rv.direction()) {
    case Direction::LEFT:
      return rv.is_push_or_pull() ? Characters::L : Characters::l;
    case Direction::RIGHT:
      return rv.is_push_or_pull() ? Characters::R : Characters::r;
    case Direction::NORTH_WEST:
      return rv.is_push_or_pull() ? Characters::U : Characters::u;
    case Direction::SOUTH_EAST:
      return rv.is_push_or_pull() ? Characters::D : Characters::d;
    case Direction::NORTH_EAST:
      return rv.is_push_or_pull() ? Characters::N : Characters::n;
    case Direction::SOUTH_WEST:
      return rv.is_push_or_pull() ? Characters::S : Characters::s;
    default:
      throw invalid_argument(
        "Illegal PusherStep direction " + to_str(rv.direction())
        + " in HexobanTessellation!"
      );
  }
}

} // namespace implementation
} // namespace sokoengine
